stream_alert/alert_merger/main.py

```diff
@@ -0,0 +1,151 @@
+
+        Args:
+            func (method): Function to invoke (ALERTS_TABLE.scan or ALERTS_TABLE.query)
+            func_kwargs (dict): Keyword arguments to pass to the scan/query function.
+                The kwargs will be modified if pagination is necessary.
+
+        Yields:
+            dict: Each item (row) from the response
+        """
+        while True:
+            response = func(**func_kwargs)
+            for item in response.get('Items', []):
+                yield item
+
+            if response.get('LastEvaluatedKey'):
+                func_kwargs['ExclusiveStartKey'] = response['LastEvaluatedKey']
+            else:
+                return
+
+    def rule_names(self):
+        """Returns the set of distinct rule names (str) found in the table."""
+        kwargs = {
+            'ProjectionExpression': 'RuleName',
+            'Select': 'SPECIFIC_ATTRIBUTES'
+        }
+        return set(item['RuleName'] for item in self._paginate(self.table.scan, kwargs))
+
+    def pending_alerts(self, rule_name, alert_proc_timeout_sec):
+        """Find all alerts for the given rule which need to be dispatched to the alert processor.
+
+        Args:
+            rule_name (str): Select all alerts from this rule name
+            alert_proc_timeout_sec (int): Alert processor timeout
+                This is used to determine whether an alert could still be in progress
+
+        Yields:
+            dict: Each alert (row) with all columns and values.
+        """
+        kwargs = {
+            # Include only those alerts which have not yet dispatched or were dispatched more than
+            # ALERT_PROCESSOR_TIMEOUT seconds ago
+            'FilterExpression': (Attr('Dispatched').not_exists() |
+                                 Attr('Dispatched').lt(int(time.time()) - alert_proc_timeout_sec)),
+            'KeyConditionExpression': Key('RuleName').eq(rule_name),
+            'Select': 'ALL_ATTRIBUTES'
+        }
+        for item in self._paginate(self.table.query, kwargs):
+            yield item
+
+    def mark_as_dispatched(self, rule_name, alert_id):
+        """Mark a specific alert as dispatched (in progress)."""
+        # Update the alerts table with the dispatch time, but only if the alert still exists.
+        # (The alert processor could have deleted the alert before the table update finishes).
+        try:
+            self.table.update_item(
+                Key={'RuleName': rule_name, 'AlertID': alert_id},
+                UpdateExpression='SET Dispatched = :now ADD Attempts :one',
+                ExpressionAttributeValues={':now': int(time.time()), ':one': 1},
+                ConditionExpression='attribute_exists(AlertID)'
+            )
+        except ClientError as error:
+            if error.response['Error']['Code'] == 'ConditionalCheckFailedException':
+                LOGGER.warn('Conditional update failed: %s', error.response['Error']['Message'])
+            else:
```

Comment:
Can you add some context as to why this particular error is okay and only logs, while others will be raised?

Reply:
It's possible (though very unlikely) that the alert processor could delete the entry from the table before the alert merger updates it with its additional state. If that happens, that's totally fine - we want the rows to be deleted when we don't need them anymore.

I removed the logging entirely - it shouldn't even be a warning, it's just an unusual order of operations. I also clarified with some more comments